### PR TITLE
feat(DMS): add resource rocketmq migration task

### PR DESCRIPTION
**Pull request by 346010762**
**What this PR does / why we need it**:
add resource RocketMQ migration task
**Special notes for your reviewer**:

**Release note**:

```release-note
add resource RocketMQ migration task
```

## PR Checklist

* [ ] Tests added/passed.
* [ ] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST=./huaweicloud/services/acceptance/dms/ TESTARGS='-run TestAccRockemqMigrationTask'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dms/ -v -run TestAccRockemqMigrationTask -timeout 360m -parallel 4
=== RUN   TestAccRockemqMigrationTask_rocketmq
=== PAUSE TestAccRockemqMigrationTask_rocketmq
=== RUN   TestAccRockemqMigrationTask_rabbitToRocket
=== PAUSE TestAccRockemqMigrationTask_rabbitToRocket
=== CONT  TestAccRockemqMigrationTask_rocketmq
=== CONT  TestAccRockemqMigrationTask_rabbitToRocket
--- PASS: TestAccRockemqMigrationTask_rabbitToRocket (717.27s)
--- PASS: TestAccRockemqMigrationTask_rocketmq (801.89s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dms       801.943s


```

### CODE REVIEW

File: docs/resources/dms_rocketmq_migration_task.md [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+
+* `exchanges` - (Optional, List, ForceNew) Specifies the exchange metadata.
+  The [exchanges](#RocketMQ_migration_task_exchanges) structure is documented below.
+  Changing this creates a new resource.
+
+* `bindings` - (Optional, List, ForceNew) Specifies the binding metadata.
+  The [bindings](#RocketMQ_migration_task_bindings) structure is documented below.
+  Changing this creates a new resource.
+
+-> **NOTE:** Parameters `vhosts`, `queues`, `exchanges` and `bindings` are set when the migration task type is **rabbitToRocket**.
```
Comment (Jason-Zhang9309) [inline docs/resources/dms_rocketmq_migration_task.md]: do you mean this?
```suggestion
-> **NOTE:** Parameters `vhosts`, `queues`, `exchanges` and `bindings` are required when the migration task type is **rabbitToRocket**.
```

File: docs/resources/dms_rocketmq_migration_task.md [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+* `write_queue_num` - (Optional, Int, ForceNew) Specifies the number of write queues. Defaults to **16**.
+  Changing this creates a new resource.
+
+<a name="RocketMQ_migration_task_subscription_groups"></a>
+The `subscription_groups` block supports:
+
+* `group_name` - (Required, String, ForceNew) Specifies the name of a consumer group.
+  Changing this creates a new resource.
+
+* `consume_broadcast_enable` - (Optional, Bool, ForceNew) Specifies whether to enable broadcast. Defaults to **true**.
```
Comment (Jason-Zhang9309) [inline docs/resources/dms_rocketmq_migration_task.md]: will the API return a `true` value when this parameter is not specified?

File: docs/resources/dms_rocketmq_migration_task.md [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+separated by a slash, e.g.
+
+```bash
+$ terraform import huaweicloud_dms_rocketmq_migration_task.test <instance_id>/<id>
+```
+
+Note that the imported state may not be identical to your resource definition, due to some attributes missing from the
+API response, security or some other reason. The missing attribute includes: `overwrite`.
+It is generally recommended running `terraform plan` after importing the resource. You can then decide
+if changes should be applied to the instance, or the resource definition should be updated to align with the instance.
```
Comment (Jason-Zhang9309) [inline docs/resources/dms_rocketmq_migration_task.md]:
```suggestion
if changes should be applied to the task, or the resource definition should be updated to align with the task.
```

File: docs/resources/dms_rocketmq_migration_task.md [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+* `topic_configs` - (Optional, List, ForceNew) Specifies the topic metadata.
+  The [topic_configs](#RocketMQ_migration_task_topic_configs) structure is documented below.
+  Changing this creates a new resource.
+
+* `subscription_groups` - (Optional, List, ForceNew) Specifies the consumer group metadata.
+  The [subscription_groups](#RocketMQ_migration_task_subscription_groups) structure is documented below.
+  Changing this creates a new resource.
+
+-> **NOTE:** Parameters `topic_configs` and `subscription_groups` are required when the migration task type is **rocketmq**.
```
Comment (houpeng80) [inline docs/resources/dms_rocketmq_migration_task.md]:
```suggestion
-> **NOTE:** Parameters `topic_configs` and `subscription_groups` are required when `type` is set to **rocketmq**.
```

File: docs/resources/dms_rocketmq_migration_task.md [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+* `exchanges` - (Optional, List, ForceNew) Specifies the exchange metadata.
+  The [exchanges](#RocketMQ_migration_task_exchanges) structure is documented below.
+  Changing this creates a new resource.
+
+* `bindings` - (Optional, List, ForceNew) Specifies the binding metadata.
+  The [bindings](#RocketMQ_migration_task_bindings) structure is documented below.
+  Changing this creates a new resource.
+
+-> **NOTE:** Parameters `vhosts`, `queues`, `exchanges` and `bindings` are required when the migration task type is **rabbitToRocket**.
```
Comment (houpeng80) [inline docs/resources/dms_rocketmq_migration_task.md]: ditto

File: huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_rocketmq_migration_task_test.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+// testRocketmqMigrationTaskImportState is used to return an import id with format <instance_id>/<id>
+func testRocketmqMigrationTaskImportState(name string) resource.ImportStateIdFunc {
+	return func(s *terraform.State) (string, error) {
+		rs, ok := s.RootModule().Resources[name]
+		if !ok {
+			return "", fmt.Errorf("resource (%s) not found: %s", name, rs)
+		}
+		instance_id := rs.Primary.Attributes["instance_id"]
```
Comment (houpeng80) [inline huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_rocketmq_migration_task_test.go]:
```suggestion
		instanceId := rs.Primary.Attributes["instance_id"]
```

File: docs/resources/dms_rocketmq_migration_task.md [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+The RocketMQ migration task can be imported using the RocketMQ instance ID and the RocketMQ migration task ID
+separated by a slash, e.g.
+
+```bash
+$ terraform import huaweicloud_dms_rocketmq_migration_task.test <instance_id>/<id>
+```
+
+Note that the imported state may not be identical to your resource definition, due to some attributes missing from the
+API response, security or some other reason. The missing attribute includes: `overwrite`.
+It is generally recommended running `terraform plan` after importing the resource. You can then decide
```
Comment (houpeng80) [inline docs/resources/dms_rocketmq_migration_task.md]:
```suggestion
It is generally recommended running `terraform plan` after importing the task. You can then decide
```

File: huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go [hunk 8]
```diff
@@ -0,0 +1,652 @@
...
+		createRocketmqMigrationTaskHttpUrl = "v2/{project_id}/instances/{instance_id}/metadata"
+		createRocketmqMigrationTaskProduct = "dms"
+	)
+	createRocketmqMigrationTaskClient, err := cfg.NewServiceClient(createRocketmqMigrationTaskProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating DMS client: %s", err)
+	}
+
+	instanceID := d.Get("instance_id").(string)
+	queryParams := fmt.Sprintf("?&name=%v&overwrite=%v&type=%v", d.Get("name"), d.Get("overwrite"), d.Get("type"))
```
Comment (houpeng80) [inline huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go]:
```suggestion
	createParams := fmt.Sprintf("?name=%v&overwrite=%v&type=%v", d.Get("name"), d.Get("overwrite"), d.Get("type"))
```

File: huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go [hunk 9]
```diff
@@ -0,0 +1,652 @@
...
+func buildCreateMigrationTaskBodyParams(d *schema.ResourceData) map[string]interface{} {
+	// RocketMQ to RocketMQ
+	if d.Get("type") == "rocketmq" {
+		topicConfigTable := map[string]interface{}{}
+		subscriptionGroupTable := map[string]interface{}{}
+		topicConfigList := d.Get("topic_configs").([]interface{})
+		subscriptionGroupList := d.Get("subscription_groups").([]interface{})
+
+		// topic metadata
+		for _, topicConfic := range topicConfigList {
```
Comment (houpeng80) [inline huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go]:
```suggestion
		for _, topicConfig := range topicConfigList {
```

File: huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go [hunk 10]
```diff
@@ -0,0 +1,652 @@
...
+	if err != nil {
+		return diag.Errorf("error creating DMS client: %s", err)
+	}
+	instanceID := d.Get("instance_id").(string)
+	deleteRocketmqMigrationTaskPath := deleteRocketmqMigrationTaskClient.Endpoint + deleteRocketmqMigrationTaskHttpUrl
+	deleteRocketmqMigrationTaskPath = strings.ReplaceAll(deleteRocketmqMigrationTaskPath, "{project_id}", deleteRocketmqMigrationTaskClient.ProjectID)
+	deleteRocketmqMigrationTaskPath = strings.ReplaceAll(deleteRocketmqMigrationTaskPath, "{instance_id}", instanceID)
+
+	deleteRocketmqMigrationTaskOpt := golangsdk.RequestOpts{KeepResponseBody: true}
+	deleteRocketmqMigrationTaskOpt.JSONBody = utils.RemoveNil(map[string]interface{}{"task_ids": []string{d.Id()}})
```
Comment (houpeng80) [inline huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go]:
```suggestion
	deleteRocketmqMigrationTaskOpt.JSONBody = map[string]interface{}{"task_ids": []string{d.Id()}}
```

File: huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go [hunk 11]
```diff
@@ -0,0 +1,652 @@
...
+}
+
+// resourceRocketmqMigrationTaskImportState is used to import an id with format <instance_id>/<id>
+func resourceRocketmqMigrationTaskImportState(_ context.Context, d *schema.ResourceData, _ interface{}) ([]*schema.ResourceData, error) {
+	parts := strings.Split(d.Id(), "/")
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid format specified for import id, must be <instance_id>/<id>")
+	}
+	d.Set("instance_id", parts[0])
+	d.SetId(parts[1])
```
Comment (houpeng80) [inline huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go]: deal the error of set method with multierror

File: huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_rocketmq_migration_task_test.go (hunk 6 above)
Comment (346010762) [reply huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_rocketmq_migration_task_test.go]: done

File: docs/resources/dms_rocketmq_migration_task.md (hunk 5 above)
Comment (346010762) [reply docs/resources/dms_rocketmq_migration_task.md]: done

File: docs/resources/dms_rocketmq_migration_task.md (hunk 4 above)
Comment (346010762) [reply docs/resources/dms_rocketmq_migration_task.md]: done

File: docs/resources/dms_rocketmq_migration_task.md (hunk 7 above)
Comment (346010762) [reply docs/resources/dms_rocketmq_migration_task.md]: done

File: huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go (hunk 8 above)
Comment (346010762) [reply huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go]: done

File: huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go (hunk 9 above)
Comment (346010762) [reply huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go]: done

File: huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go (hunk 10 above)
Comment (346010762) [reply huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go]: done

File: huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go (hunk 11 above)
Comment (346010762) [reply huaweicloud/services/dms/resource_huaweicloud_dms_rocketmq_migration_task.go]: done

**Pull request merged**